src/if_statements.py

```diff
@@ -654,36 +629,29 @@ def build_flowgraph_between(
             curr_start = fallthrough
             continue
 
-        # Switch nodes are always marked emit_goto.
-        assert not isinstance(curr_start, SwitchNode)
-
-        if isinstance(curr_start, BasicNode):
-            # In a BasicNode, the successor is the next articulation node.
-            curr_start = curr_start.successor
-        elif isinstance(curr_start, ConditionalNode):
-            # A ConditionalNode means we need to find the next articulation
-            # node. This means we need to find the "immediate postdominator"
-            # of the current node, where "postdominator" means we have to go
-            # through it, and "immediate" means we aren't skipping any.
-            curr_end = immediate_postdominator(context, curr_start, end)
-            # We also need to handle the if-else block here; this does the
-            # outputting of the subgraph between curr_start and the next
-            # articulation node.
-            body.add_if_else(
-                build_conditional_subgraph(context, curr_start, curr_end, indent)
-            )
-            # Move on.
-            curr_start = curr_end
-        else:  # ReturnNode
-            # Write the return node, and break, because there is nothing more
-            # to process.
-            add_return_statement(context, body, curr_start, indent, last=False)
-            break
+        # The interval to process is [curr_start, curr_start.immediate_postdominator)
+        curr_end = curr_start.immediate_postdominator
+        assert curr_end is not None
+
+        # For nodes with branches, curr_end is not a direct successor of curr_start
+        if isinstance(curr_start, ConditionalNode):
+            body.add_if_else(build_conditional_subgraph(context, curr_start, curr_end))
+        elif isinstance(curr_start, SwitchNode):
+            body.add_switch(build_switch_between(context, curr_start, curr_end))
+        else:
+            # No branch, but check that we didn't skip any nodes
+            if curr_start.children() != {curr_end}:
+                raise DecompFailure(
+                    f"While emitting flowgraph between {start.name()}:{end.name()}, "
+                    f"skipped nodes while stepping from {curr_start.name()} to {curr_end.name()}."
+                )
```

Comment:
Should probably be a warning, if it can occur legitimately. Some output is better than none.

Reply:
This shouldn't occur legitimately if our calculation of `immediate_postdominator` isn't buggy, I just wanted to include enough info to help debug the problem - should I maybe update the comment? change to an assert?

Reply:
an assert-with-error-message is the precedent in the codebase I think

Reply:
Yeah you're right, "this should be impossible" is better left as an assert I think